### DISCUSSION TITLE
Constituents information

### DISCUSSION THREAD

**Discussion by shubhani16** [MadAnalysis/madanalysis5#212] [Q&A]
Hi

Hope you are doing well.

I want to store charged and neutral constituents of the jet in the madanalysis5 expert mode. Is it possible to separate out the constituents of the jet based on separately using the charged and neutral constituents for further analysis?

Any help on this would be much appreciated. Thanks.

Regards
Shubhani

**Comment by jackaraz** [comment]
Hi @shubhani16, my answer will depend on what you need to do. In v1.10, every jet object includes all the constituent information, but this is a pointer to the MC event, so you can, in principle, access all the information you need about the constituents. You can find more information in [this reference card](https://github.com/MadAnalysis/madanalysis5/tree/main/doc/expert_mode). But if you need more sophisticated information, I will suggest looking into v2. There you can even implement constituent-level detector simulation, and all the jet constituents will also include that information. For that, you can look into [arXiv:2303.03427](https://arxiv.org/abs/2303.03427) [hep-ph].

Hope it helps
Cheers

**Reply by shubhani16** [reply]
Hi Jack

Thanks for your reply. My current toolkit consists of grabbing eflow tracks, eflow hadrons, and photons and then using expert mode to cluster the particles into jets using Fastjet. From, there I want to store the charged constituents and neutral constituents separately. I can access the constituents using a pointer as you said:
for(MAuint32 i=0;i<Leading_BJet[0].constituents().size();i++)
      {
        myfile2 << i+1 << "," << Leading_BJet[0].constituents()[i].user_index() << "," << Leading_BJet[0].constituents()[i].pt() << "," <<
                  Leading_BJet[0].constituents()[i].m() << "," << Leading_BJet[0].constituents()[i].eta() << "," << Leading_BJet[0].constituents()[i].phi()<< "\n";
      }

My question was whether we have a pointer in madanalysis that can be used to check whether the constituent is charged or neutral and then store that information.

Regards
Shubhani

**Comment by jackaraz** [comment]
Hi @shubhani16, you can extract the charge information via [GetCharge function](https://github.com/MadAnalysis/madanalysis5/blob/f96f336d1dfd4733b8dde26976782d5294994744/tools/SampleAnalyzer/Commons/Service/PDGService.h#L86) 

PS: you will need to add 
```c++
#include "SampleAnalyzer/Commons/Service/PDGService.h"
```
to your code then you can use ```PDG->GetCharge(...)```

**Reply by shubhani16** [reply]
Hi Jack

Thanks for your reply.

This solves my problem. Thanks.

Regards
Shubhani

**Reply by shubhani16** [reply]
Hi Jack

One more question on this front. I am using fastjet pseudojet and then grabbing its constituents. Using the PDG->GetCharge(...) gives me error: invalid initialization of reference of type 'const MA5::MCParticleFormat&' from expression of type '__gnu_cxx::__alloc_traits<std::allocator<fastjet::PseudoJet>, fastjet::PseudoJet>::value_type' {aka 'fastjet::PseudoJet'}

To grab constituents I am using: sub_Jets[i].constituents()[j].

Is there any other way I can initialize to grab the charge?

Regards
Shubhani

**Reply by jackaraz** [reply]
Hi @shubhani16, nothing is designed to work with pseudojet in v1.10 for that you need to go to v2, so you either need to provide MCParticleFormat or pdgid.

**Reply by shubhani16** [reply]
Hi @jackaraz, thanks will try v2.

**Reply by jackaraz** [reply]
GetCharge function still does not accept pseudojet in v2 either, its designed for madanalysis particle objects not anything else. Pseudojet object does not necessarily hold the information of the particle charge or identity its simply a fourvector. V2 is designed to do jet substructure analyses and fastjet is used under the hood not by the user.

**Reply by shubhani16** [reply]
Would it be possible to grab the constituents pdgid and then use the GetCharge function?

**Reply by jackaraz** [reply]
yes, you can simply check the ```GetCharge``` function

https://github.com/MadAnalysis/madanalysis5/blob/f96f336d1dfd4733b8dde26976782d5294994744/tools/SampleAnalyzer/Commons/Service/PDGService.h#L85-L102

To see what types it accepts.

**Reply by shubhani16** [reply]
thanks, will have a look.
